java/flight/flight-core/src/main/java/org/apache/arrow/flight/auth2/BasicAuthValidator.java

```diff
@@ -0,0 +1,194 @@
+   */
+  interface CredentialValidator {
+    /**
+     * Validate the supplied credentials (username/password) and return the peer identity.
+     * @param username The username to validate.
+     * @param password The password to validate.
+     * @return The peer identity if the supplied credentials are valid.
+     * @throws Exception If the supplied credentials are not valid.
+     */
+    String validateCredentials(String username, String password) throws Exception;
+  }
+
+  /**
+   * Interface that this validator delegates for generating and validating a token.
+   */
+  interface AuthTokenManager {
+    /**
+     * Generate a token for the supplied username and password.
+     * @param username The username to generate the token for.
+     * @param password The password to generate the token for.
+     * @return The generated token.
+     */
+    String generateToken(String username, String password);
+
+    /**
+     * Validate the supplied token and return the peer identity.
+     * @param token The token to be validated.
+     * @return The peer identity if the supplied token is valid.
+     * @throws Exception If the supplied token is not valid.
+     */
+    String validateToken(String token) throws Exception;
+  }
+
+  @Override
+  public AuthResult validateIncomingHeaders(CallHeaders incomingHeaders) throws Exception {
+    final AuthResult authResult = parseAndValidateBearerHeaders(incomingHeaders);
+    if (authResult != null) {
+      return authResult;
+    }
+    return parseAndValidateBasicHeaders(incomingHeaders);
+  }
+
+  /**
+   * Parses and validates basic headers and returns the AuthResult.
+   * @param incomingHeaders Incoming header to parse.
+   * @return An instance of an AuthResult.
+   * @throws Exception When the basic header does not contain the credentials or the credential validation fails.
+   */
+  private AuthResult parseAndValidateBasicHeaders(CallHeaders incomingHeaders) throws Exception {
+    final String authEncoded = AuthUtilities.getValueFromAuthHeader(incomingHeaders, Auth2Constants.BASIC_PREFIX);
+    if (authEncoded == null) {
+      throw CallStatus.UNAUTHENTICATED.toRuntimeException();
+    }
+    // The value has the format Base64(<username>:<password>)
+    final String authDecoded = new String(Base64.getDecoder().decode(authEncoded), StandardCharsets.UTF_8);
+    final int colonPos = authDecoded.indexOf(':');
+    if (colonPos == -1) {
+      throw CallStatus.UNAUTHENTICATED.toRuntimeException();
+    }
+
+    final String user = authDecoded.substring(0, colonPos);
+    final String password = authDecoded.substring(colonPos + 1);
+    final String peerIdentity = credentialValidator.validateCredentials(user, password);
+    return new AuthResult() {
+      @Override
+      public String getPeerIdentity() {
+        return peerIdentity;
+      }
+
+      @Override
+      public CallHeaderAuthenticator.HeaderMetadata getHeaderMetadata() {
```

Comment:
I don't see why we need this method. It looks like it caches information to use later in appendToOutgoingHeaders(), but you can just as easily use this information directly since getHeaderMetadata() and appendToOutgoingHeaders() are implemented at the same time.

I also see this being used in registerBearer which I'll comment on separately.

Reply:
Removed the interface and the method.